WIP: supporting remote MCP servers from the web.

csells opened the pull request:
**tl;dr:**

```sh
dart pub global run pana .
## ✓ Follow Dart file conventions (30 / 30)
## ✓ Provide documentation (20 / 20)
## ✓ Platform support (20 / 20)
### [*] 20/20 points: Supports 6 of 6 possible platforms (iOS, Android, Web, Windows, macOS, Linux)
### [*] 0/0 points: WASM compatibility

This package is compatible with runtime `wasm`, and will be rewarded additional points in a future version of the scoring model.

## ✓ Pass static analysis (50 / 50)
## ✓ Support up-to-date dependencies (40 / 40)

Points: 160/160.
```

# CHANGELOG
## 0.6.0
- Added web browser and WASM compatibility for MCP clients via StreamableHttpClientTransport
- Add edFlutter client demo showcasing cross-platform MCP client usage
- Implemented conditional compilation for platform-specific optimizations while maintaining API consistency
    - the existing StreamableHttpClientTransport for non-web platforms has been kept for it's performance and reliability.
    - the new StreamableHttpClientTransport is for web and still being tuned

# NOTES
- no use of EventFlux -- the AI decided to go along way
- the Zapier MCP server endpoint didn't work due to CORS, but huggingface and deepwiki both do. I wonder what's going to happen in the future with that
- deepwiki access is sssllllooooowwww. I assume that because it's the one with session management.

# TESTING
- Apparently the VSCode testing support doesn't do browser-hosted tests, but read the TESTING__GUIDE.md for instructions for running all of the tests
- And more details in WEB_CLIENT_TESTS_SUMMARY.md

# TECHNICAL DESIGN
The AI likes to brag, but if you're curious, check out:
- WEB_DESIGN_DOC.md

leehack commented on README.md:
```diff
@@ -1,39 +1,58 @@
+  identical APIs
+- **MCP Servers** (Dart VM only):
+  - Stdio support (Server and Client)
+  - StreamableHTTP support (Server only)
```
It should be server and client.

leehack commented on README.md:
```diff
@@ -1,39 +1,58 @@
+  - StreamableHTTP support (Server only)
+  - SSE support (Server only) - Deprecated  
+  - Stream Transport using dart streams (Server and Client in shared process)
+- **MCP Clients** (Cross-platform):
+  - Stdio support (VM only)
+  - **StreamableHTTP support** - **Works everywhere: VM, web browsers, WASM!**
```
I feel it's redundant. Maybe we don't need to split Client/Server in this section.

leehack commented on README.md:
````diff
@@ -84,9 +103,59 @@ void main() async {
 }
 ```
 
+### Client (Cross-Platform: VM, Web, WASM)
````
```suggestion
### Client
```

csells commented:
@leehack you're the expert here -- I was just working with an agent on this one. I was pleased with the web and wasm compat and the ability to pass the tests and connect to real-world remote MCP servers. You obviously know more than I do about the specifics.

And the "dart vm" thing was just how the AI chose to designate web vs. non-web. You're right about AOT of course.

leehack commented:
@csells Thanks for the PR. Wow. It's a huge PR. I started a review, but it will take some time for sure. Btw, haven't you noticed any issues with the notifications? (I'll be testing it myself later tho)

I see your(or AI's) approach is similar to what I've tried (https://github.com/leehack/mcp_dart/pull/32)

The issue I had was that, weirdly, if I make more than 5 requests, it starts pending. I didn't have enough time to debug all of them, but I had the impression that every request tries to initiate the SSE connection, and it finally hits Chrome's limit. Everything was still fine on mobile tho.

csells commented:
> I've briefly tested, but it seems the implementation doesn't work with SSE properly.

yep. I'm working on that now along with a perf fix for deepwiki.

csells commented:
take a look and see if the SSE stuff is better. also, there may be a fix to your Fixed "More than 5 requests pending" bug. Here's what the AI reports:

Root Cause: Single reusable HTTP client hit browser connection limits (6 per domain), causing requests >5 to become permanently pending

Solution: Replaced single client with connection pool of 4 HTTP clients using round-robin distribution to stay under browser limits

Testing: Added comprehensive test suite (web_multiple_requests_test.dart) that reproduces the issue and validates the fix with 10+ concurrent requests

Results:
✅ 10+ concurrent requests complete successfully (~152ms)
✅ No more permanently pending requests
✅ All existing functionality preserved
✅ Production-ready and backward compatible

Key Changes:
Connection pool with _getPooledClient() round-robin selection
Enhanced timeout configuration with browser-optimized presets
Improved SSE handling and reconnection logic

leehack commented:
@csells I'll be merging https://github.com/leehack/mcp_dart/pull/32 .
I tested it with other streamable HTTP servers, and it is working fine. I still see the 5th request pending issue with my demo mcp server. I'll still be debugging the issue, but with the merge, the mcp web client should be working.

leehack commented:
Closing it in favor of https://github.com/leehack/mcp_dart/pull/32
